File: src/leetcodejava/tree/BinaryTree102.java
package leetcodejava.tree;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

/**
 * 二叉树的层次遍历
 *
 * @author: zhangyu
 */
public class BinaryTree102 {

    /**
     * 二叉树层次遍历
     *
     * @param root 根节点
     * @return 返回遍历链表
     */
    public List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer>> res = new ArrayList();
        levelHelper(res, root, 0);
        return res;
    }

    /**
     * 递归遍历
     *
     * @param res    结果链表
     * @param root   根节点
     * @param height 高度
     */
    public void levelHelper(List<List<Integer>> res, TreeNode root, int height) {
        if (root == null) return;
        if (height >= res.size()) {
            res.add(new LinkedList<>());
        }
        res.get(height).add(root.val);
        levelHelper(res, root.left, height + 1);
        levelHelper(res, root.right, height + 1);
    }

    /**
     * 二叉树层次遍历
     *
     * @param root 根节点
     * @return 返回遍历链表
     */
    public List<List<Integer>> levelOrder2(TreeNode root) {
        List<List<Integer>> res = new ArrayList<>();
        Queue<TreeNode> queue = new LinkedList<>();
        queue.add(root);
        while (!queue.isEmpty()) {
            List<Integer> list = new ArrayList();
            int size = queue.size();
            for (int i = 0; i < size; i++) {
                TreeNode p = queue.poll();
                list.add(p.val);
                if (p.left != null) {
                    queue.add(p.left);
                }
                if (p.right != null) {
                    queue.add(p.right);
                }
            }
            res.add(list);
        }
        return res;
    }
}

File: src/leetcodepython/tree/binary_tree_102.py
'''
二叉树层次遍历
author:zhangyu
date:2020.2.14
'''
from typing import List

from tree.tree_node import TreeNode


def level_helper(res: List[List[int]], root: TreeNode, height: int):
    '''
        循环遍历二叉树
    Args:
        res: 链表
        root: 根节点
        height: 高度
    '''
    if root == None:
        return
    if height >= len(res):
        return res.append([])
    res[height].append(root.val)
    level_helper(res, root.left, height + 1)
    level_helper(res, root.right, height + 1)


def level_order1(root: TreeNode) -> List[List[int]]:
    '''
        二叉树层次遍历
    Args:
        root: 跟节点
    Returns:
        层次遍历返回值
    '''
    res = []
    level_helper(res, root, 0);
    return res


def level_order2(root: TreeNode) -> List[List[int]]:
    '''
        找出抢劫最大值
    Args:
        root: 跟节点
    Returns:
        最大值
    '''
    res = []
    queue = []
    queue.append(root)
    while len(queue) > 0:
        arr = []
        size = len(queue)
        for i in range(size):
            p = queue.pop()
            arr.append(p.val)
            if p.left != None:
                queue.append(p.left)
            if p.right != None:
                queue.append(p.right)
        res.append(arr)
